File: src/arrray/scala/PeakIndexInMountain.scala
package arrray.scala

object PeakIndexInMountain {

  def peakIndexInMountainArray(arr: Array[Int]): Int = {
    binarySearch(arr, 0, arr.length - 1)
  }

  def binarySearch(arr: Array[Int], min: Int, max: Int): Int = {
    val mid = ((min + max) / 2).floor.toInt
    if (min == max) min
    else if (arr(mid) > arr(mid + 1) && arr(mid) > arr(mid - 1)) mid
    else {
      if (arr(mid) > arr(mid - 1)) binarySearch(arr, mid + 1, max)
      else binarySearch(arr, mid, mid - 1)
    }
  }

  //  def binarySearch(arr: Array[Int], min: Int, max: Int): Int = {
  //    var index = 0
  //    while (min <= max) {
  //      var min = 0
  //      var max = arr.length - 1
  //      val mid = ((min + max) / 2).floor.toInt
  //      if (arr(mid) > arr(mid + 1) && arr(mid) > arr(mid - 1)) index = mid
  //      else {
  //        if (arr(mid) > arr(mid - 1)) max = mid - 1
  //        else min = mid + 1
  //      }
  //    }
  //    index
  //  }

  def main(args: Array[String]): Unit = {
    val arr = Array(0, 2, 1, 0)
    val arr01 = Array(3, 4, 5, 1)
    val arr02 = Array(3, 5, 3, 2, 0)
    val rs = peakIndexInMountainArray(arr02)
    println(rs)
  }
}

File: src/arrray/java/PeakIndexInMountain.java
package arrray.java;

public class PeakIndexInMountain {

    public static int peakIndexInMountainArray(int[] arr) {
        return binarySearch(arr, 0, arr.length - 1);
    }

    public static int binarySearch(int[] arr, int min, int max) {
        int mid = ((min + max) / 2);
        if (min == max) return min;
        else if (arr[mid] > arr[mid - 1] && arr[mid] > arr[mid + 1]) return mid;
        else {
            if (arr[mid] > arr[mid - 1]) return binarySearch(arr, mid + 1, max);
            else return binarySearch(arr, mid, mid - 1);
        }
    }


    public static void main(String[] args) {
        int[] input = {0, 2, 1, 0};
        int[] input01 = {3, 5, 3, 2, 0};
        int rs = peakIndexInMountainArray(input01);
        System.out.println(rs);
    }
}
